testapp/testapp/AppDelegate.swift

```diff
@@ -0,0 +1,156 @@
+//
+//  AppDelegate.swift
+//  testapp
+//
+//  Created by Christopher Batin on 16/01/2019.
+//  Copyright © 2019 Christopher Batin. All rights reserved.
+//
+
+import UIKit
+import Ably
+import UserNotifications
+
+// 1
+let authURL = "YOUR_NGROK_HTTPS/auth"
+@UIApplicationMain
+class AppDelegate: UIResponder, UIApplicationDelegate, ARTPushRegistererDelegate {
+    var window: UIWindow?
+    var realtime: ARTRealtime!
+    var channel: ARTRealtimeChannel!
+    var subscribed = false
+
+    //2
+    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+        DispatchQueue.main.async() {
+            print("** didRegisterForRemoteNotificationsWithDeviceToken")
+            ARTPush.didRegisterForRemoteNotifications(withDeviceToken: deviceToken, realtime: self.getAblyRealtime())
+        }
+    }
+
+    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
+        DispatchQueue.main.async() {
+            print("** didFailToRegisterForRemoteNotificationsWithError")
+            ARTPush.didFailToRegisterForRemoteNotificationsWithError(error, realtime: self.getAblyRealtime())
+        }
+    }
+
+    //3
+    private func getAblyRealtime() -> ARTRealtime {
+        let options = ARTClientOptions()
+        options.authCallback = { params, callback in
+            self.getTokenRequest() { json, error in
+                do {
+                    callback(try ARTTokenRequest.fromJson(json!), nil)
+                } catch let error as NSError {
+                    callback(nil, error)
+                }
+            }
+        }
+        realtime = ARTRealtime(options: options)
```

Comment:
Be careful, `self.getAblyRealtime` is called twice so the `realtime` object will change.

Reply:
That's why your syncing with `DispatchQueue.main`?

Reply:
It ensures that ably realtime can't nil. I think for this use case it's okay to call it in multiple locations as I've not created a singleton